### DISCUSSION TITLE
Questions: How should issues be managed, how to run etc

### DISCUSSION THREAD

**Discussion by slashgordon42** [helge17/tuxguitar#14] [General]
@helge17: As I understand, this project is a fork of the SVN dir at 

https://sourceforge.net/projects/tuxguitar/files/TuxGuitar/TuxGuitar-1.5/

This has a list of tickets (Bugs, Feature requests)

I found some bugs, e.g. 
- musicxml export creates files which lead to an error message when importing them with musescore
- transposing a guitar line -12 half notes leads to loss of many of the notes
 My questions: 
- should these things be issued as tickets in the sourceforge repo? Or as issues in this GitHub repo?
- Is there a way of adding a "Discussion" to this repo?
- is there a description how to setup/compile this project locally?

**Comment by guiv42** [comment]
Thanks @helge17 for opening the "Discussions"
just one part of the answer: you can refer to [this example script](https://github.com/helge17/tuxguitar/blob/master/misc/build_tuxguitar_from_source.sh). It needs to be adapted to your environment, but it makes really clear the different steps to compile locally

**Reply by slashgordon42** [reply]
@helge17: thanx for opening the Discussions
@guiv42: thanx for your reply. I had already found this script and had adapted it for my machine. Basically, I commented out 

[ `uname` == Linux ] && start_remote_*

so only

[ `uname` == Linux ] && build_tg_for_linux

should be exec'd on my box. 

However, building fails with several errors, starting with

     [exec] gcc -I/home/wolfgang/projects/git-cos/tuxguitar/build-scripts/native-modules/tuxguitar-alsa-linux-x86_64/../common-include -fPIC -c -o org_herac_tuxguitar_player_impl_midiport_alsa_MidiSystem.o org_herac_tuxguitar_player_impl_midiport_alsa_MidiSystem.c
     [exec] org_herac_tuxguitar_player_impl_midiport_alsa_MidiSystem.c:3:10: fatal error: alsa/asoundlib.h: Datei oder Verzeichnis nicht gefunden
     [exec]     3 | #include <alsa/asoundlib.h>
     [exec]       |          ^~~~~~~~~~~~~~~~~~
     [exec] compilation terminated.
     [exec] make: *** [GNUmakefile:13: org_herac_tuxguitar_player_impl_midiport_alsa_MidiSystem.o] Fehler 1


Looks like dependencies are missing. 

Any hints welcome. I attached the output of the run. 

[compile_out.txt](https://github.com/helge17/tuxguitar/files/11256294/compile_out.txt)

**Reply by guiv42** [reply]
Yes, looks like you're missing a dependency. On my system I had to add the following packages to build the native-modules:
- alsa-devel (for tuxguitar-alsa-linux-x86_64)
- libjack-devel (for tuxguitar-jack-linux-x86_64)
- fluidsynth-devel (for tuxguitar-fluidsynth-linux-x86_64)
- libliv-0-devel, suil-devel (for tuxguitar-synth-lv2-linux-x86_64)

**Reply by slashgordon42** [reply]
Ah, thanx, was able to some of the gcc issues with
sudo apt-get install libasound2-dev
and
sudo apt install libjack-jackd2-dev

fluidsynth-devel I did not find, trying

sudo apt install libfluidsynth-dev

Still can't find a package for suil-devel  (Ubuntu)

**Reply by slashgordon42** [reply]
Seems to be
liblilv-dev
and
libsuil-dev

**Reply by helge17** [reply]
Maybe I can include some patches from Sourceforge in this repository here, but it's much better if you open new issues here on Github, or even better, pull requests.

**Comment by guiv42** [comment]
I don't know ubuntu, neither the package search tools it may provide... On my side I could identify the required packages by looking at the missing .h header file in the build error, then I searched for packages providing it

18 avr. 2023 00:58:51 slashgordon42 ***@***.***>:

**Reply by slashgordon42** [reply]
Yeah, I used a package search engine:

https://packages.ubuntu.com

Now it's missing Qt, will search on another day. Thanx for your support!

**Reply by guiv42** [reply]
I don't really understand why Qt is needed if you use the provided build script (which I think doesn't build the Qt version). Building Qt version is a bit more [tricky](https://github.com/helge17/tuxguitar/blob/master/build-scripts/tuxguitar-linux-qt5-x86_64/todo.txt).
What are your objectives exactly? If you just need one working build in your configuration, you don't need the full build script, you don't even need to build/install a deb package.
```
cd build-scripts/tuxguitar-linux-swt-x86_64
mvn clean verify -P native-modules  
./target/tuxguitar-SNAPSHOT-linux-swt-x86_64/tuxguitar.sh
```
should be sufficient (should be OK for  swt or jfx)

**Reply by slashgordon42** [reply]
Installed the QT libraries with

sudo apt-get install qt5-default qt5-qmake qtbase5-dev-tools qt5-doc 

Then tried what you proposed. Got a little bit further. Fails now with

     [exec] g++ -I/home/wolfgang/projects/git-cos/tuxguitar/build-scripts/native-modules/tuxguitar-synth-vst-linux-x86_64/include -fPIC -fpermissive -c -o VSTClient.o VSTClient.c
     [exec] In file included from VSTClient.h:4,
     [exec]                  from VSTClient.c:7:
     [exec] VST.h:4:10: fatal error: aeffectx.h: Datei oder Verzeichnis nicht gefunden
     [exec]     4 | #include <aeffectx.h>
     [exec]       |          ^~~~~~~~~~~~
     [exec] compilation terminated.
     [exec] make: *** [GNUmakefile:43: VSTClient.o] Fehler 1

Have to find out where to get aeffectx.h

**Reply by guiv42** [reply]
The answer is in the [example build script](https://github.com/helge17/tuxguitar/blob/master/misc/build_tuxguitar_from_source.sh)
See `Copy header files of the Steinberg SDK (VST_SDK_2.4) in place ...`
You need to copy the files here: build-scripts/native-modules/tuxguitar-synth-vst-linux-x86_64/include/

**Reply by slashgordon42** [reply]
Ah, thanx for the hint, I'll try this.

**Reply by slashgordon42** [reply]
Thanx you very very much. This was the last bit, I just startet my first self-compiled version! Awsesome!
@guiv42, @helge17: I wonder if it's worthy to write this down in a simple text file for lazy noobs like me?

Could be something like

"Howto-build-locally.txt"

with sections for different OSs?

So I might write a description for Ubuntu. Could be plain text, rst, or md also I don't really like it.

**Reply by guiv42** [reply]
Why not... Somehow it will be partially redundant with the example build script, and there is always a risk it may not be maintained. But this discussion shows it may be helpful.
In fact the only information that is really missing is the list of dependencies required to build, with different package names for different distros.
By the way, if you want to describe the *simplest* path to a local build on a Linux x86-64 machine, I think you should refer to the jfx version. At least that removes the SWT installation issue #1 I initially had.

**Comment by guiv42** [comment]
For Debian users:
I had a try in a Debian docker image, and with the following list of packages installed I could successfully build (manually) the jfx and swt versions, and generate .deb and .rpm packages:
`rpm binutils file maven make gcc libasound2-dev libjack-jackd2-dev libfluidsynth-dev liblilv-dev libsuil-dev g++ qtbase5-dev fakeroot`
The resulting container is about 1Go, against the ~120Mo of the "bare" Debian official Docker image

**Comment by helge17** [comment]
There is this [INSTALL.md](https://github.com/helge17/tuxguitar-2.0beta/blob/master/INSTALL.md) file in the tuxguitar-2.0beta version, that's a good starting point I think. I could copy it here.

**Reply by slashgordon42** [reply]
Ah, that's kewl. Exactly what I was looking for, did not know it does exist.
